Source/WebCore/dom/Element.cpp

```diff
@@ -560,18 +560,37 @@ Ref<Node> Element::cloneNodeInternal(Document& targetDocument, CloningOperation
 {
     switch (type) {
     case CloningOperation::OnlySelf:
-    case CloningOperation::SelfWithTemplateContent:
         return cloneElementWithoutChildren(targetDocument);
+    case CloningOperation::SelfWithTemplateContent: {
+        Ref clone = cloneElementWithoutChildren(targetDocument);
+        ScriptDisallowedScope::EventAllowedScope eventAllowedScope { clone };
+        cloneShadowTreeIfPossible(clone);
+        return clone;
+    }
     case CloningOperation::Everything:
         break;
     }
     return cloneElementWithChildren(targetDocument);
 }
 
+void Element::cloneShadowTreeIfPossible(Element& newHost)
+{
+    RefPtr oldShadowRoot = this->shadowRoot();
+    if (!oldShadowRoot || !oldShadowRoot->isCloneable())
+        return;
+
+    Ref clone = static_cast<Node&>(*oldShadowRoot).cloneNodeInternal(newHost.document(), Node::CloningOperation::SelfWithTemplateContent);
```

Comment:
We should make cloneNodeInternal() public in ShadowRoot instead of doing this cast. There is no good reason for this function to be public on Node and private on ShadowRoot AFAICT. Also, it prevents de-virtualization of the function call.

Reply:
That's a good point. Will do.